### PR TITLE
[IMP] contract_invoice_merge_by_partner: Don't require to merge invoices later

### PR DESCRIPTION
**Pull request by pedrobaeza**
Adapting source `contract` module, we don't need to rely on account_invoice_merge functionality for having all the invoices merged.

cc @Tecnativa

### CODE REVIEW

**Comment by pedrobaeza** [comment]
cc @carlosdauden @cubells

**Review by cubells — APPROVED** [review]
:+1: 

LGTM

Test and code review.

Two minor things only:

- you must edit installation section in README, and remove accordingly that module account_invoice_merge is necessary

- in previous version, origin field in the new invoice created, contained the names of the contracts merged, in your version, only first contract name appears in origin field.

**Comment by pedrobaeza** [comment]
Thanks for the review. I have squashed the fix for the proposed changes as there's little benefit in 2 big diffs for the same thing.

**Comment by yajo** [comment]
This one has conflicts.

**Comment by pedrobaeza** [comment]
Conflicts solved. You can review the PR and let's get finally merged.

File: contract/__openerp__.py [hunk 1]
```diff
@@ -1,11 +1,12 @@
 # -*- coding: utf-8 -*-
 # © 2004-2010 OpenERP SA
 # © 2016 Carlos Dauden <carlos.dauden@tecnativa.com>
+# © 2015-2017 Pedro M. Baeza <pedro.baeza@tecnativa.com>
```
Comment (yajo) [inline contract/__openerp__.py]: https://github.com/OCA/maintainer-tools/pull/197

File: contract/models/contract.py [hunk 2]
```diff
@@ -1,17 +1,16 @@
 # -*- coding: utf-8 -*-
 # © 2004-2010 OpenERP SA
 # © 2014 Angel Moya <angel.moya@domatix.com>
-# © 2015 Pedro M. Baeza <pedro.baeza@tecnativa.com>
+# © 2015-2017 Pedro M. Baeza <pedro.baeza@tecnativa.com>
```
Comment (yajo) [inline contract/models/contract.py]: https://github.com/OCA/maintainer-tools/pull/197

File: contract/models/contract.py [hunk 3]
```diff
@@ -222,72 +221,103 @@ def _prepare_invoice_line(self, line, invoice_id):
...
     def _prepare_invoice(self):
-        self.ensure_one()
-        if not self.partner_id:
+        """Prepare the values for the invoice creation from the contract(s)
+        given. It's possible to provide several contracts. Only one invoice
+        will be created and most of the values will be taken from first
+        contract, but there are certain values that can be obtained from all
+        of them (for example, the origin field).
+
+        :param: self: Recordset of contract(s).
```
Comment (yajo) [inline contract/models/contract.py]: It's `:param self:`.

http://www.sphinx-doc.org/en/stable/domains.html?highlight=param#python-signatures

File: contract/models/contract.py [hunk 4]
```diff
@@ -222,72 +221,103 @@ def _prepare_invoice_line(self, line, invoice_id):
...
             ctx.update({
                 'old_date': old_date,
                 'next_date': new_date,
-                # Force company for correct evaluate domain access rules
+                # For correct evaluating of domain access rules + properties
                 'force_company': contract.company_id.id,
             })
-            # Re-read contract with correct company
-            invoices |= contract.with_context(ctx)._create_invoice()
+            obj = self.with_context(ctx)
```
Comment (yajo) [inline contract/models/contract.py]: In general I think it's easier to use kwargs, which include the previous context and just override the supplied keys.

    obj = self.with_context(old_date=old_date, new_date=new_date, force_company=contract.company_id.id)

The result is the same, so not a requirement 😸

**Review by yajo — APPROVED** [review]
Just some minor details

**Comment by pedrobaeza** [comment]
Well, that changes don't concern the PR focus itself, but I have satisfied your desire, hehe.

**Comment by pedrobaeza** [comment]
As everything is satisfied, I merge this for not delaying more the patch.

**Pull request merged**